### DISCUSSION TITLE
Any interest in my "man in the middle" solution for Sec+ 1.0 and 0x37 wall control?

### DISCUSSION THREAD

**Discussion by mcomer-captech** [Kaldek/rat-ratgdo#95] [Ideas]
Hey, checking to see if there might be any interest in a solution I have up and running for a problem with older GDOs. Specifically, I have a Sec+ 1.0 "purple button" opener and an older wall control which came with it. It turns out that this wall control is of the 0x37 variety - that is, it does not query door status in the same way other controls do (with a 0x38 command). The folks over in the RATGDO code-base are trying to solve this by injecting a 0x38 into the command stream when they detect a 0x37 wall panel - but for whatever reason this does not work for me, something about the timing is not quite right and my opener is not responding to the 0x38 when it is sent like this (though it does respond when this is done in wall control emulation).

So - I built a work-around. I added a 12v source and 2 additional MOSFETS so that my project is now man-in-the-middle between the GDO and wall control. My board does wall control emulation for the GDO and GDO emulation for the wall control. I relay commands from the wall control to the opener (toggle light, toggle door, toggle lock) and relay status from the opener to the wall control. I have this up and running with the ESPHome code-base from the RATGDO folks, somewhat modified to add the separate interface to the wall control.

Right now I have this all soldered up on a prototype board but am thinking of taking the KiCAD schematics and PCB files from here, modifying, and ordering something more "professional".

Would be happy to contribute back if there is any interest? My use-case may be a little niche, I don't know. Not sure how many folks have older GDOs (Sec+ 1.0) and 0x37 wall controls.

**Comment by TheCranston** [comment]
@mcomer-captech I'm very interested.  It's been in the back of my mind that the proverbial "man-in-the-middle" behaviour would solve several issues at once.  Please count me in.  Happy to help with code or hardware as I can.

**Comment by foreverimagining** [comment]
Definitely very interesting. And nowhere near as niche as you'd expect.

I don't have a Security+1.0 GDO myself, but there are a lot of people in the HA forums that do and have problems with it, as well as people with "dumb" or smart-but-not-compatible wall panels that don't necessarily want to have to replace them. I wonder if this would be the solution for them, and it could probably be made as an add-on board to original-flavor RATGDOs (though I suspect some soldering would be necessary).

Do you ever see any of the reported issues like the GDO opening by itself at random, general command unreliability, or the lock/light buttons on the wall panel not functioning? I'm pretty sure there have been more issues with Security+1.0, but those are the ones I remember off the top of my head.

**Comment by mcomer-captech** [comment]
Ok, thanks for the feedback, sounds like I should create some schematics and a board layout for what I have done and should share. I also have my code additions already available online, but not ready to link to it yet as it does need some cleanup. (Doesn't code always?)

Another couple of questions then @TheCranston  @foreverimagining: 

1. Right now I have a little 12v "wall wart" power supply providing the extra 12v source needed for the wall control. My thinking is that I will feed this from my opener's battery backup instead - and while I am at it, will add a little 5v regulator to power the microcontroller. That would make the entire setup power-fail safe as the opener, controller, and power to the wall control would all be coming from the opener's backup. Thoughts?
2. I am currently using a D1 Mini Lite in my setup - and so to add the extra serial comms for the wall control I had to sacrifice a couple of pins. There are not enough pins available for 2x serial to the GDO, 2x serial to the wall control, 3x dry contact in (open, close, light), and 2x status outputs (door, obstruction). To keep everything I'd need a device with more pins exposed. Should I just design around a device with more available GPIOs? (It didn't matter to me since I don't need the dry contact or status output features - but I guess you never know who might want them.)

**Comment by mcomer-captech** [comment]
> Do you ever see any of the reported issues like the GDO opening by itself at random, general command unreliability, or the lock/light buttons on the wall panel not functioning? I'm pretty sure there have been more issues with Security+1.0, but those are the ones I remember off the top of my head.

I have only had my solution actually installed and running for about 2 days, but so far I have seen none of those issues. Wall control button presses work great, there have been no unexpected openings or closings, etc. 

The only odd glitch that I see every once in a while - and I saw this PRIOR to adding my extra wall-control comms - is a little "bounce" in the light toggle command from ESPHome. Occasionally I will hit "toggle light" (while it is off) and the UI in Home Assistant toggles to on, toggles back to off, then toggles back to on (where it is supposed to be) - all within a second or two. I think I know what is going on there and I think it is a fundamental problem in how the opener sends status, how that gets reported back to ESPHome / HomeAssistant, etc. Essentially - there is an ongoing update of light and lock status (either the wall controller or the emulation code is sending a 0x3a to request status) and then that status has to get propagated up to HA via the API. This means that there is latency in that update, so an "old" status can come in even after the command to toggle has been sent. It's a UI glitch only and it sorts itself out quickly, so...

**Comment by foreverimagining** [comment]
> Another couple of questions then

1. Having the option to run it off a 12v wall supply or the battery backup would probably be best. There seem to be a fair few people who have come to this repository that have battery backups and want to make use of them, but I've always thought not having one was more the norm for most GDOs, so having the choice would benefit everyone. You can probably check out the other schematics/boards in this repository to see how they've done it if you want to compare.
As for the microcontroller, correct me if I'm wrong, but they actually run on 3.3V and the boards have a voltage regulator to step the 5V from USB down to the required 3.3V. The RATGDO boards also have a 3.3V IN pin broken out for an alternative to USB power (though I don't know how many people take that option since everyone probably has a few USB cables/bricks lying around). It would make more sense, then, to use 3.3V connected to the 3.3V pin, unless someone has a specific need for 5V like the ultrasonic sensor I'm adding to mine.

2. When it comes to Security+ GDOs, I expect most people keep to the TX, RX, Blk/Obst side of the board and ignore the dry contact/status output side, which is meant to create greater compatibility with other types of GDOs but otherwise is largely unnecessary. The exception to that might be, for instance, gate openers, but they have a lot of problems to contend with getting RATGDO to work in general. I would say keep the scope smaller to start with and drop the dry contact side. Dry contact support could be added later as a separate, dedicated board or a larger, combined board if there are people who need both.
To support people with Security+1.0 and original RATGDOs, it wouldn't be too far outside of that same, smaller scope to make another board with just the additional middleman components. Since there are unnecessary components connected between the pins and the breakouts, though, I think the options will be either to create a three board stack, RATGDO/Middleman/D1 Mini, or to solder wires directly to the pins on the ESP8266 module to have access to them. Unfortunately, because of regular design changes that we in the 3D printing community abhor, the first option will only work with v2.5, and the second option would be necessary for v2.5x's.



> a little "bounce"

If this is what I think it is, I've seen it happen on various switches/lights in Home Assistant and would also consider it to be a mild buffering-type annoyance rather than a possible issue with the device. Which means so far so good. Long term reliability can really only be determined over time, but sounds very promising so far.

**Comment by mcomer-captech** [comment]
Quick update on this: I wanted to let my solution "bake" a little bit before I moved forward with anything. I can happily report that in a couple of weeks of testing now, I have had absolutely zero issues. No unexplained door openings. No odd behavior from the wall control buttons, controls in Home Assistant, controls in Apple Home (I expose the door and related items via the Homebridge integration in HA). The motion sensor on the wall control turns on the light and that is immediately reflected in HA. Etc. I am very happy with this.

I do have some time away scheduled for later this week so I will be away from home for a couple of weeks. Once I am back my plan is to:
- Produce a schematic and pcb design in KiCad, have some prototype boards made, etc.
- Update my software solution to include the latest from the Sec+1.0 work on the RATGDO side (my current code is based on what they had a few days prior to their merge to master so I need to make a few updates.)

In the meantime - if anyone is super-interested in this, absolutely needs a solution for the Sec+1.0 opener and 0x37 wall control, etc - ping me and I can hopefully share more details prior to getting all of the polishing-up done.

**Reply by windsurferwill** [reply]
I'd like to replicate your implementation. Would you be so kind to share your details?
Steve

**Comment by jrsphoto** [comment]
I think I also have an older purple button opener but not sure what opener I have.  My opener looks like [this](https://www.amazon.com/Liftmaster-41A5273-1-Multi-Function-Control-Panel/dp/B000F5KF1O/ref=sr_1_1?crid=1TQNLZNRNZGP4&keywords=LiftMaster+security%2B+1+Control+Panel&qid=1707022534&sprefix=liftmaster+security%2B+1+control+panel%2Caps%2C121&sr=8-1).  Just wondering your your version is what I'm looking for.

**Comment by ur7x** [comment]
Count me in too... Post your schematic when you are ready... I have a couple of Security 1+ Red button openers with 889LM wall panels and  I would be happy to build one and test it... 

Presently I'm running MQTT on the original schematic here. With one week of use my  experience is:
1) The RATGDO software will report a MQTT "Syncing" status a couple of times a day.  "Covers" in HA do not understand what syncing is and this seems to result in phantom door close commands and light toggle commands.  "Syncing" status events last about 30 seconds and during this time the doors (and lights) are unresponsive. 
2) The light control in RATGDO is a toggle (now confirmed) so if the light and the light's status get out of sync an "ON" command from HA will turn off lights that are already on.  
3) So far only one phantom garage door open...  But that is one too many. 

Happy with the local solution and happy with local light control (even if it is flaky) It is ironic that MyQ forced this solution upon us and the resulting "hack" already has more functionality than the cloud system.

**Comment by windsurferwill** [comment]
I am interested.  I have a security 1 GDO (purple) with a 398.  I can’t switch to a 889 control panel because it’s too different and harder to use functions like lock out. But the door status with the 398 doesn’t work.  Man in the middle sounds like a great solution.  What can I do to help? Oh, I have the latest ratgdo v2.52 running in MQTT mode with HA. Works flawlessly except the cover doesn’t report door status.

**Comment by mitchjs** [comment]
im jumping in here, i did work (just for fun) on the HOMEKIT version of ratgdo...
i added 1.0 support
while doing this... i learned alot... 
and i totally rewrote how SEC+1.0 works and know why there are issues... 
i believe my code for SEC+1.0 is the most reliable one to date

its about knowing when you can "TALK" on the buss, and how to avoid bit collisions.

i only have a 889lm, and worked with that... not sure what a 398 is
if you have an iphone/ipad and want to try my firmware out
(why its still sitting as a PULL REQUEST, i dont know)

the MQTT fw could get cleaned up, not a fan cause i dont feel like running mqtt broker
(but i have one running for testing this stuff)
the esphome code is well written...
i worked on the HOMEKIT version for a friend... as i really didnt need it for my situation
but its rocking...

i think i got it nailed, i have had zero false openings, and 100% accurate status
the only time i see collision is with light, and its rare... and my code does its best to trap
the garbled byte out... so within fw 100ms status gets correct on next poll by 889lm

@mcomer-captech 
i think the man-in-the middle could solve 100% of the knowing when to talk issues
right now i use timing, when the byte is transmitted from the lm889 and the GDO byte is sent back
and i then i know i have a "window" to send my byte out (if i have one to send)
i also have code to detect collision garbage and ignore
you can look at my source in comms.cpp
and i be happy to work with anyone
is your schematic for ESP32 or the esp8266
i think the code would work in either
me fan of esp32 variants

i do electronics too, and was thinking of producing a ESP32 based one myself
but since i haven't had real need for this personally... i guess been lazy
(my 3rd party control still works on HE, dont know why but it does)


Mitch

**Comment by windsurferwill** [comment]
I'm currently using MQTT in Home Assistant with RATGDO, but I could use ESPHome if it worked with my Security1+ GDO.  From reading the latest posts, apparently the reason my GDO is not providing open/close status is my 398LM wall control unit uses the 0x37 message.  I cannot switch to the newer 889LM wall controller which uses 0x38 messages.  I have confirmed that the RATGDO properly provides open/close status when I disconnect my 398LM to put RATDO into emulation mode which tells me the garage door opener itself is capable of open/close status.  

@mcomer-captech , others: 
Could the man-in-the-middle approach solve this? Does anyone know if the 0x37 wall control requires a response to it's message?  Could the man-in-the-middle intercept the 0x37 and respond with 0x38?
Steve

**Comment by mitchjs** [comment]
Im sure i could fix any of the FW to work with 398LM, but other than this panel being old... why cant you switch to newer 889LM? if i had access to 398LM (and it worked with my gdo... which prob does)

0x38 is query.
0x37 could also be a query.
as you see I have 2 unknowns (i just never seen the messages with my 889LM so that's why i call em  unknown)

i was thinking of just turning the Homekit firmware, in to a direct http control...
so one can make a simple Hubitat driver (as that is what i got)
MQTT not a fan of having another device as a broker
ESPHOME version of course is good for that ecosystem


   ```
 DoorButtonPress = 0x30,
 DoorButtonRelease = 0x31,
 LightButtonPress = 0x32,
 LightButtonRelease = 0x33,
 LockButtonPress = 0x34,
 LockButtonRelease = 0x35,

 Unkown_0x36  = 0x36,
 Unknown_0x37 = 0x37,

DoorStatus = 0x38,
ObstructionStatus = 0x39,   // this is not proven
LightLockStatus = 0x3A,
```

**Comment by foreverimagining** [comment]
> why cant you switch to newer 889LM?

889LM's aren't exactly cheap to simply buy as a replacement, and despite being regarded as compatible and even recommended, they are not always problem-free. While your technique may be able to fix the code to solve some or all of those problems, some people also have wall panels with more capabilities, such as a thermometer, and they don't want to have to downgrade to an 889LM, even if the cost were easily justifiable. This man-in-the-middle option will hopefully give people options that will allow them to choose the best course of action for themselves. Though, if you are able to fix the code for 0x37 wall panels, I'm sure there are a lot of people who would be appreciative.

**Comment by mitchjs** [comment]
looking on ebay, 398LM isnt cheap!!
if i can get one, someone loan me one... i can work on it

**Comment by windsurferwill** [comment]
I really appreciate you jumping in. I wish I had one to loan you but I can't take mine out of service.  Would it help if I could run experiments, dumps, etc.?

**Comment by mitchjs** [comment]
@windsurferwill, you using the ESPHOME version of ratgdo? got an iPhone? could flash what I did for homekit
and you could test on iPhone (dont need anything more than iPhone as homekit app is on it)
and can add some code to log more (and i got logging in there)
since i have no idea what 0x36/0x37 messages are, since for one i never seen one on my system
889LM+GDO

**Comment by windsurferwill** [comment]
@mitchjs: No sorry, I do not have an iPhone. I am using the MQTT version with RATGDO v2.52 hardware because from the comments on Paul's RATGDO GitHub page ESPHOME is still a work-in-progress for Security 1+ openers.

**Comment by mitchjs** [comment]
that code is problematic too... i started with that code and made MANY improvements in the sec1.0 code. but only for Homekit version. not sure if i have time to play with mqtt version right now... i also dont like that i need to have a broker running on some box, and have my Hubitat...

**Comment by mitchjs** [comment]
i see paul made some improvements in latest fw (2.57) make sure you at that

**Comment by ur7x** [comment]
> i see paul made some improvements in latest fw (2.57) make sure you at that

Thanks for that, ya that works a lot better and ends HA turning on the lights and then the motion sensor toggling them off as I walk into the garage. 

The 889LM's sure do sync a lot.. I have read elsewhere that they have internal "super capacitors" that are used to power the wall unit like a battery... How often should they be syncing?  Maybe one of my capacitors has gone bad.

**Comment by mitchjs** [comment]
what do you mean "sync" where the lights on the unit flash... that should only happen at power up
if i remember it was the 888LM with the bad caps... and 889LM was the replacement, i might be wrong
i dont remember if any 889LMs had issues

**Comment by ur7x** [comment]
HA reports the door status as "syncing" which is unknown for a "cover" and the wall units flash with a yellow and red light behind the door button.  Does this for about 20-30 seconds  a couple of times a day... While this happens the wall unit is unresponsive. 

Just doubled checked.. You were right.. they are 888LM's date coded from Nov 2016.. Is that in the range for Cap replacements?

**Comment by mitchjs** [comment]
yes, thats not good... should not happen.... maybe wiring... maybe bad pannel
and its an 889?

**Comment by ur7x** [comment]
> yes, thats not good... should not happen.... maybe wiring... maybe bad pannel and its an 889?

Just doubled checked.. You were right.. they are 888LM's date coded from Nov 2016.. Is that in the range for Cap replacements?

**Comment by mitchjs** [comment]
maybe, check liftmaster web site site
if not, its 2 1uf 2.7v caps

**Comment by mitchjs** [comment]
https://cloud.info.liftmaster.com/888LMReplacement

**Comment by ur7x** [comment]
> https://cloud.info.liftmaster.com/888LMReplacement

In true Liftmaster style.  They email me a coupon for a "Free 889LM"  Yeah!... When I go to use it...Only the shipping is free... They want  $58 dollars for the wall unit.  Salt meet wound. 

I've ordered the Caps to solder in.

**Comment by mitchjs** [comment]
@ur7x, a friend of mine says, call them, they might swap it for no cost!!
i believe thats what he did...

**Comment by Sudo-Rob** [comment]
Add me to the list of the interested.  Please consider my [alternate approach](https://github.com/Kaldek/rat-ratgdo/pull/75) to the GDO and wall panel interfaces.

**Comment by windsurferwill** [comment]
@mcomer-captech
I want to try your circuitry and code.  Maybe I missed it, but could you share?  My scenario is exactly the one you originally posted. Thanks in advance!

**Reply by mcomer-captech** [reply]
Hi everyone - I need to apologize - I posted this and have offered to share, but have been terribly slow in getting anything done in terms of posting. I tend to get myself involved in too many projects both at work as well as personally and I think I am just at saturation point right now. I am going to try and do better!

Quick update: this has continued to work flawlessly for me since I installed it. I have had zero problems - no unexplained door openings, no failures to open or to control the light or anything else when commanded, my older Sec+ 1.0 0x37 wall control works just fine, etc.

Notes:
- The circuitry is quite simple and just an extension of what is already posted here. All I did was add another couple of MOSFETS, a couple of resistors, and a little 12v wall adapter to power the wall control. These new components provide independent power to the wall control (the GDO normally does this but of course we're keeping the wall control separate from the GDO here) and provide a separate communication channel (using sw serial) to the microcontroller.
- My software solution is based on the ESPHome fork of RATGDO but could be adapted to other stuff. I finished my work up right before the main Sec+ 1.0 work was merged into the RATGDO codebase so my code might be touch out of date vs. the main codebase - not sure if it is worth me fixing that or not. I preferred the native ESPHome integration method so that's what I went with and what I extended.
- I am essentially just using the RATGDO EPSHome code to talk to the GDO - and then bolted on additional code to trick the wall control into thinking it is directly connected to a GDO when it isn't.
- My solution could certainly be simplified and improved - like for example building in a little voltage down-converter so that wall control power and microcontroller power could come from the same source, making a provision for powering both from the GDO backup battery (which my GDO has but not all do I guess), re-baselining against the latest RATGDO ESPHome repo, etc. I just have not had the time to do any of that yet.

Seems like priorities would be for me to a) get a schematic posted, and b) get link to my software posted - so I'll try and make that happen.

**Reply by mbelling** [reply]
@mcomer-captech - I'm also interested in this.  Do you have the code online anywhere, even pre-rebaseline?

**Reply by windsurferwill** [reply]
Not trying to pressure but is there anything you can share on your code or schematics?  I didn't see anything in your repositories.  Love to help further the man in the middle solution.
Steve

**Reply by mcomer-captech** [reply]
Ok, my fault - I know that I opened this conversation and then never followed-through. Work busy, home busy, projects busy, etc - you guys know how it is.

Here is a link to my sources where this is implemented and working for me. I'll see if I can do something quickly for a schematic, but it is pretty straightforward - I just replicated the input buffer and driver from the GDO side to create a separate "channel" to talk to the wall control and powered it with a little 12vdc wall brick.

https://gitea.va-comer.net/mcomer/esphome-ratgdo/src/branch/comer

The code there is WAAAAY behind the main-line now, and I am exposing some of my dirty laundry in commit-history as I worked through figuring out how custom ESPHOME components worked and debugged all of this. Let me know if this helps!

**Reply by mcomer-captech** [reply]
And I just realized that my external DNS is not resolving to the right IP right now so the link above is not working. Should have that sorted shortly, sorry!

**Reply by windsurferwill** [reply]
mcomer - Given the evolution of the main ESPhome-ratgdo  (do I have it right that message contention has improved?), would your opinion today be that it still makes sense to utilize man-in-the-middle?
I am still running a security 1 GDO (purple) with a 398 and ratgdo v.252 in MQTT mode.  I have to keep the 398 as the primary door panel.

**Comment by mitchjs** [comment]
just a FYI, i worked on this project; https://github.com/ratgdo/homekit-ratgdo. 
I only added SEC1.0 to it, as i did it for a friend and don't use ratgdo currently
(because my direct hubitat method still oddly works)
anyway, i solved 99% of the data collisions without the man in the middle (and added some code to reject any collided data), I watched the timing of the polling of the 889LM
I put in rules to not transmit, until proper time (in milliseconds)
I do think Man in the Middle is a REALLY good idea
does anyone know if one is allowed to wire up more than 1 889LM to a GDO?
if so, clearly Liftmaster knows when its ok to transmit
I wonder if both poll... or one goes quiet and only listens
(just thinking out loud)
havent looked at this still in some time... but find it very interesting :)

**Comment by mitchjs** [comment]
@mcomer-captech , question for you, i see you mention 0x37 type wall panel... what wall panel you got?
i got a 889LM and never seen it

**Comment by soul350** [comment]
@mitchjs I just went down the RATGDO rabbit hole, proto'd a board to work with my purple button Security 1+ and am about finished designing a custom board to be the permanent install for it.  I have a 78LM wall panel, and the light button no longer works on it (probably the lock button too, but I have never used that).  I was about to dig in and see if I could modify the code to fix the non-working light button when I found this thread.  I'm currently running the main RATGDO code via ESPHome.  Are your changes compatible with that?  

I also thought about making a man in the middle solution, but if your approach fixes it the majority of the time, that's enough for me.

**Comment by mitchjs** [comment]
@soul350 , I am involved with the homekit-ratgdo project here on github (both esp8266 and esp32)
specifically, the SEC+1.0 (since that's my GDO) and i got it working FLAWLESS with a little non-public "add on" to the ratgdo product
and no man-in-the-middle needed, which means no need to deal with power for the digital wall plates.

I'm not involved with the esphome FW

now you have a 78LM (dirt cheap to replace btw) this is what we call a dumb panel, and the RatGDO will do emulation of an 889LM panel to make it all work, since your panel isn't going to be colliding messages, I would think the ESPHome firmware just work fine
and pretty sure it has been improved over time... load it up, test it out

also a FYI i dont use the apple homekit at all, the homekit-firmware has a really nice http rest interface, which I use in my own Hubitat driver .
keeps everything simple for me.

**Reply by soul350** [reply]
How does the 78LM panel communicate the light button press?  I assume the door opener button just pulls the line to ground with no signalling, since it works fine.  The light (and lock) must send some comms though.  I am running the latest ESPHome FW on mine.  the Security 1+ flavor.

Semi-related, do the openers with a purple button but red/white/black screw terminals instead of the typical RWWB quick release terminal block also support the Security 1+ protocol, or are they dry contact compatible only?

**Reply by soul350** [reply]
After replying, I remembered that my scope was still in the garage.  Pardon the horrible signal integrity, this is connected to a 10' length of wire so I can work on a bench under the opener.  It looks like a press of the light button is signaled by changing the width of the pulses on the control line.  (This is with no RATGDO attached)
<img width="800" height="480" alt="DS1Z_QuickPrint30" src="https://github.com/user-attachments/assets/826bf127-1b77-403e-a911-dce76c2c096c" />

Zoom:
<img width="800" height="480" alt="DS1Z_QuickPrint31" src="https://github.com/user-attachments/assets/6d9e9d5d-9c48-461a-b859-270dc93b6fac" />

**Comment by mitchjs** [comment]
@soul350 

the 78LM uses resistors and nothing more.... (never had one, but pretty sure, would love to see PCB!)
id suspect you can pop it off the wall, and pop the pcb off the back and see

as for:
Semi-related, do the openers with a purple button but red/white/black screw terminals instead of the typical RWWB quick release terminal block also support the Security 1+ protocol, or are they dry contact compatible only?
i don't know... mine is from 2006, so it has the RWWB block... but scope shows pulses, id say yes the SEC+1.0 wire protocol is in play

one thing I do know, with NO DIGITIAL panel, the GDO will put out a pulse stream (attempting to sync a digitial panel)
so that is what you are seeing...

**Reply by soul350** [reply]
I found images someone else took, along with a partial description, and made this.  Looks like the open button just shorts it, and the lock and light buttons put caps across the lines to stretch the pulses like I saw on my scope.
<img width="772" height="1160" alt="image" src="https://github.com/user-attachments/assets/4156d809-dd57-4349-9aed-398944dbd380" />

**Comment by mitchjs** [comment]
@soul350 

Nice, thanks for answering my curiosity!!!!

yea.. so they must have the buss on a pin and measure the pulse time 
i might have to nab a 78LM just to test against
(i got some 10uf and 100uf caps, i might test out and see if it causes any issues with the UART connection in ratdgo)

what kind of "ratgdo" did you build esp8266 or (hopefully) ESP32?
id love for you to flash this firmware and test 

ESP8266: https://ratgdo.github.io/homekit-ratgdo/flash.html
ESP32: https://ratgdo.github.io/homekit-ratgdo32/flash.html

assuming you did the same PINS as ratdgo, if not i can make you custom firmware :)



mitch

**Reply by soul350** [reply]
I did an ESP32-S3 via ESPHome with custom pins.  I think the answer is that my light is just not going to work unless I upgrade to a newer wall switch, which honestly isn't worth it to me for something I use so rarely.  I could also wire it directly to the ESP32 instead of the GDO and write some custom code to monitor it, but again, probably not worth the effort.  The bulb was burned out in my GDO for like 6 months and I didn't even care lol.

**Reply by mitchjs** [reply]
so it should work, its not?

yea... i have some code to monitor the wire, i did it for ES8266 *cause i had extra, and just need mosfet
but skipping that for now

it should work... for sure without the panel connected, but really with it too

**Reply by soul350** [reply]
The light button on the wall panel doesn't work, but the door opener does, that's good enough.  I suspect that once the RATGDO starts panel emulation, the GDO stops looking for stretched pulses.  That or the polling messes with the stretched pulses.

**Reply by mitchjs** [reply]
yes.... once the polling starts, there is no more pulse stream
just a poll every 250ms
even man in the middle wouldnt work
so.. option #1 378LM, option #2 889LM
there are 3rd party 378LMs pretty cheap (its a wireless panel, that you learn in)
https://www.ebay.com/itm/266510800113

**Comment by mitchjs** [comment]
@soul350 , oh in the end what system you going to be controlling it from... 
I use Hubitat :)

**Reply by soul350** [reply]
Home Assistant

**Comment by mitchjs** [comment]
@soul350 
this was all known i gues...
from ratgdo page

Analog wall panels (such as the 78LM) do not communicate with the door opener at all, and in this case ratgdo is able to query the garage door opener for status directly. This allows ratgdo to know the door state, but it has the side effect of disabling the analog wall panel’s light and lock buttons (the door open button still works normally).

so its for sure one of the options i mentioned :)

**Comment by mitchjs** [comment]
@soul350 , there should be an input pin on the design that when grounded will tell ratgdo to turn on light
you might not have a wire available in your run from the GDO to that 78LM
(who ever did mine, ran CAT5 so i have 4 pairs!)
but you could "mod" that 78lm and get light feature working
(or just nab a 378LM)
